src/vmm/src/devices/legacy/serial.rs

```diff
@@ -277,6 +284,24 @@ impl<I: Read + AsRawFd + Send + Debug> MutEventSubscriber
     }
 }
 
+/// Checks whether the given file descriptor is a FIFO pipe.
+fn is_fifo(fd: RawFd) -> bool {
+    let mut stat = std::mem::MaybeUninit::<libc::stat>::uninit();
+
+    // SAFETY: No unsafety can be introduced by passing in an invalid file descriptor to fstat,
+    // it will return -1 and set errno to EBADF. The pointer passed to fstat is valid for writing
+    // a libc::stat structure.
+    if unsafe { libc::fstat(fd, stat.as_mut_ptr()) } < 0 {
+        return false;
+    }
+
+    // SAFETY: We can safely assume the libc::stat structure to be initialized, as libc::fstat
+    // returning 0 guarantees that the memory is now initialized with the requested file metadata.
+    let stat = unsafe { stat.assume_init() };
+
+    (stat.st_mode & libc::S_IFIFO) != 0
+}
```

Comment:
Nit: I like to wrap `std::mem::MaybeUninit::uninit`  in the `unsafe` block so outside `unsafe` we don't need to deal wiht it.

This is a stylistic choice so feel free to ignore.

```rust
/// Checks whether the given file descriptor is a FIFO pipe.
fn is_fifo(fd: RawFd) -> bool {
    // SAFETY: We can safely assume the libc::stat structure to be initialized, as libc::fstat
    // returning 0 guarantees that the memory is now initialized with the requested file metadata.
    let stat = unsafe {
        let mut stat = std::mem::MaybeUninit::<libc::stat>::uninit();
        // No unsafety can be introduced by passing in an invalid file descriptor to fstat,
        // it will return -1 and set errno to EBADF. The pointer passed to fstat is valid for writing
        // a libc::stat structure.
        if libc::fstat(fd, stat.as_mut_ptr()) < 0 {
            return false;
        }
        stat.assume_init()
    };
    (stat.st_mode & libc::S_IFIFO) != 0
}
```

Reply:
Mh, I like having each "unsafe-causing" call be its own block, although using scoping to make sure that the MaybeUninit does not escape is indeed nice! I'll keep it like this to not need to regather approvals, but thanks for the idea!